apis/python/src/tiledbsoma/options/_soma_tiledb_context.py

```diff
@@ -101,11 +215,23 @@ def replace(
 
             >>> context.replace(tiledb_config={"vfs.s3.region": "us-east-2"})
         """
-        if tiledb_config:
-            new_config = self.tiledb_ctx.config()
-            new_config.update(tiledb_config)
-            changes["tiledb_ctx"] = tiledb.Ctx(config=new_config)
-        return attrs.evolve(self, **changes)
+        with self._lock:
+            if tiledb_config is not None:
+                if tiledb_ctx:
+                    raise ValueError(
+                        "Either tiledb_config or tiledb_ctx may be provided"
+                        " to replace(), but not both."
+                    )
+                new_config = dict()
+                new_config = self._internal_tiledb_config()
+                new_config.update(tiledb_config)
```

Comment:
I'm not sure if this is intentional, but....
* if I pass `tiledb_config` to `__init__`, the user-provided config entirely replaces the default config
* if I pass `tiledb_config` to `replace`, it simply updates the existing config

I can see cases for both, but I am wondering if they should not be consistent?

Also, how do I remove a config option using `replace`?

Reply:
In both cases this would be equivalent, right?

- When passing the config to `__init__`, there is (almost) no existing config to replace. It will overwrite any entry of the `_default_config` that overlaps (technically, the one that overlaps).
- When passing it to `replace`, it does the same thing, but the base configuration is that of the existing context object.

> Also, how do I remove a config option using `replace`?

This wasn’t supported in the existing API, but passing `None` as a value seems like as good an idea as any.

Reply:
> When passing the config to __init__, there is (almost) no existing config to replace. It will overwrite any entry of the _default_config that overlaps (technically, the one that overlaps).

With `__init__`, this approach will not work for any key in the default config, but not specified by the user.

I had assumed that the default config was "best practices", and that we would want to retain this config for any context where the user does not specifically overwrite it. Consider a case where the user is only creating a context because they want to set the AWS S3 region - in this case, I argue they want to inherit the default configuration and simply update (add to it, i.e., add the S3 region config param).

If this isn't true, then I'm unclear why we ever have a default configuration in the code...

So, assuming we have defaults for a good reason, then I think we need to follow one of the following options:

a) Keep code as is, but add public access to the default config, so that I can do:

      > ctx = SOMATileDBContext(SOMATileDBContext.default_config.update({'vfs.s3.region': 'us-west-1'}))

b) Or, have `__init__` do the above by default (i.e., use the merge of the default config and user-provided config), and accept keys with a `None` value when I want to nuke them.

In either case, supporting the "delete `None` valued keys" seems smart.